rust/phoenix-channel/src/lib.rs

```diff
@@ -638,10 +679,28 @@ enum OkReply<T> {
 #[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
 #[serde(rename_all = "snake_case")]
 enum ErrorInfo {
-    Reason(String),
-    Offline,
+    Reason(Reason),
+    Offline, // TODO: Why is this not part of `KnownError`?
+    Disabled,
```

Comment:
@conectado I don't think I understand the structure here fully. `Offline` seems to be sent for offline resources? Why are `Offline` and `Disabled` out here and then `Reason` has `KnownError` which are also variants?

Can we unify the error reporting here to a common structure?

cc @AndrewDryga

Reply:
Also, `Offline` seems to be a _domain-specific_ error (only sent as a response for certain messages) so I think it should actually go into `R` type parameter. IMO, the `phoenix-channel` library shouldn't know about domain-specific errors from particular messages.

Reply:
phoenix-channel errors are not very consistent, and sometimes you get a reason, sometimes you get an outside error. The way I have organized the types around it is how I found best to react to them and design strong-typing around them, the format was found heuristically since there's no spec(or the javascript client is the spec?).

Feel free to change it, I have wanted to clean this up for a long time.

Reply:
JS client is the spec :)

Reply:
> `offline` is an error that is returned when you try to access a resource that doesn't have online relays/gateways, `disabled` is an error that can be returned only when you try to create a log sync and this feature is disabled. So they are unified, but you probably need to expect them to differ depending on which message was sent.

Would it be possible to send `offline` and `disabled` also as a nested struct with a `reason`?

Reply:
How you want them to look? Do we need extra structure just to wrap an enum?

Reply:
It would be nice if errors would always look the same. Currently, we have the following structures (see the difference in `payload`):

### Payload with `status` and `response`
```json
{"event":"phx_reply","ref":1,"topic":"client","payload":{"status":"error","response":"disabled"}}
```

### Payload with `reason`
```json
{"event": "disconnect", "ref": null, "topic": "client", "payload": { "reason": "token_expired" }}
```

### Payload with `status `and `response` with nested `reason`
```json
{"event": "phx_reply", "ref": null, "topic": "client", "payload": { "status":"error", "response":{ "reason":"unmatched topic" }}}
```

Reply:
Aligning the `disconnect` event isn't as critical, that can stay separate.

But aligning the format across all `phx_reply` messages would be nice!

Reply:
This one is not generated by us but we can make other errors look like this one structure-wise
```
{"event": "phx_reply", "ref": null, "topic": "client", "payload": { "status":"error", "response":{ "reason":"unmatched topic" }}}
```
but this one is harder because it's not a reply, so `event` will be different even if we keep the structure similar:
```
{"event": "disconnect", "ref": null, "topic": "client", "payload": { "reason": "token_expired" }}
```
I think for `event=disconnect` you should just ignore the payload, not even sure what I would send in `payload.status` here as it is not an error.

Reply:
I've opened a PR to make general error replies look like this:
`{"event": "phx_reply", "ref": null, "topic": "client", "payload": { "status":"error", "response":{ "reason":"unmatched topic" }}}`